zerver/lib/bulk_create.py

```diff
@@ -145,3 +157,124 @@ def bulk_create_streams(realm: Realm,
     Recipient.objects.bulk_create(recipients_to_create)
 
     bulk_set_users_or_streams_recipient_fields(Stream, streams_to_create, recipients_to_create)
+
+def bulk_create_reactions(
+    messages: Iterable[Message] = (),
+    users: Iterable[Any] = (),
+    emojis: Iterable[Any] = ()
+) -> None:
+    messages = list(messages)
+    if not emojis:
+        emojis = _DEFAULT_EMOJIS
+    emojis = list(emojis)
+
+    reactions: List[Reaction] = []
+    for message in messages:
+        reactions.extend(_create_random_reactions_to_message(
+            message, emojis, users))
+    Reaction.objects.bulk_create(reactions)
+
+
+_DEFAULT_EMOJIS = [
+    ('+1', '1f44d'),
+    ('smiley', '1f603'),
+    ('eyes', '1f440'),
+    ('crying_cat_face', '1f63f'),
+    ('arrow_up', '2b06'),
+    ('confetti_ball', '1f38a'),
+    ('hundred_points', '1f4af'),
+]
+
+
+def _create_random_reactions_to_message(
+        message: Message,
+        emojis: Iterable[Any],
+        users: Optional[Iterable[Any]] = None,
+        prob_reaction: float = 0.075,
+        prob_upvote: float = 0.5,
+        prob_repeat: float = 0.5) -> List[Reaction]:
+    '''Randomly add emoji reactions to each message from a list.
+
+    Algorithm:
+
+    Give the message at least one reaction with probability `prob_reaction`.
+    Once the first reaction is added, have another user upvote it with probability
+    `prob_upvote`, provided there is another recipient of the message left to upvote.
+    Repeat the process for a different emoji with probability `prob_repeat`.
+
+    If the number of emojis or users is small, there is a chance the above process
+    will produce multiple reactions with the same user and emoji, so group the
+    reactions by emoji code and user profile and then return one reaction from
+    each group.
+    '''
+    for p in (prob_reaction, prob_repeat, prob_upvote):
+        if p > 1 or p < 0:
+            raise ValueError('Probability argument must be between 0 and 1.')
+    EMOJI_NAME = 0
+    EMOJI_CODE = 1
+    emojis = itertools.cycle(emojis)
+    if not users:
+        if message.recipient.type == Recipient.PERSONAL:
+            users = [
+                UserProfile.objects.get(recipient__id=message.recipient.id),
+                message.sender]
+        else:
+            users = [
+                subscription.user_profile
+                for subscription in Subscription.objects.filter(
+                    recipient=message.recipient)]
+        if not users:
+            return []
+    users = itertools.cycle(users)
```

Comment:
This might result in a deterministic order; we really want to be using `random.choice()` rather than `next()` iteration, I think.

Reply:
We also ideally don't want to do an O(users) operation before we've even decided if we're doing any work.  Not sure what the right way to do this is.

Reply:
If you want, I can add an optimization to potentially avoid performing database queries for users if there will be no reactions to begin with:

```python
def _add_random_reactions_to_message(...):
    ...
    compute_next_reaction: bool = random.random() < prob_reaction
    if not compute_next_reaction:
        return []
    if not users:
          # database queries for users
          ...
     emojis = list(emojis)
     reactions = []
     while compute_next_reaction:
         ...
         compute_next_reaction = random.random() < prob_reaction

```

Reply:
I think that'd be well worth doing, assuming my theory is correct that profiling will show this is a significant regression.  

Oh, one simplification -- I bet you can just use `UserMessage.objects.filter(message=message).values_list("user_profile_id", flat=True)` for the `users` query rather than the if/else clause you have with Subscription; that will just return the IDs and might be enough to make such a change unnecessary.

But best to profile first.